### PR TITLE
secret-create passes through the rotate duration

### PR DESCRIPTION
**Pull request by wallyworld**
secret-create passes through the rotate duration and it is stored in metadata
The scope parameter is removed as it is not used.

Add backend state support for updating secrets - a future PR will wire up the update hook command and api.

Also tweak the list-secrets output.

## QA steps

bootstrap and deploy an ubuntu unit

```
$ juju exec --unit ubuntu/0 "secret-create --rotate 1h apitoken devil666"
secret://v1/7d9de3a3-c231-4667-8e3f-d953484b430c/58a9cf14-f926-4c00-8fa5-d4e130e94bdb/ubuntu.apitoken
$ juju exec --unit ubuntu/0 "secret-create password secret!"
secret://v1/7d9de3a3-c231-4667-8e3f-d953484b430c/58a9cf14-f926-4c00-8fa5-d4e130e94bdb/ubuntu.password

$ juju secrets
ID  Revision  Rotate  Backend  Path             Age
1          1  1 hour  juju     ubuntu.apitoken  7 minutes ago  
2          1  never   juju     ubuntu.password  3 seconds ago
```

### CODE REVIEW

**Comment by wallyworld** [comment]
[create secret stores rotate duration](https://trello.com/c/luCGms19/1725-create-secret-stores-rotate-duration)

File: cmd/juju/secrets/list.go [hunk 1]
```diff
@@ -159,6 +162,42 @@ func (c *listSecretsCommand) Run(ctxt *cmd.Context) error {
 	return c.out.Write(ctxt, details)
 }
 
+func humanizeDuration(duration time.Duration) string {
```
Comment (hpidcock) [inline cmd/juju/secrets/list.go:165]: This is nice, probably be good to extract to a utill package. But I think generally single character suffixes is sufficient.

`1d 5h 25m 59s`  is better than  `1 day 5 hours 25 minutes 59 seconds` which is quiet long.

File: cmd/juju/secrets/list_test.go [hunk 2]
```diff
@@ -47,17 +49,22 @@ func (s *ListSuite) TestListTabular(c *gc.C) {
...
 	s.secretsAPI.EXPECT().Close().Return(nil)
 
 	ctx, err := cmdtesting.RunCommand(c, secrets.NewListCommandForTest(s.store, s.secretsAPI))
 	c.Assert(err, jc.ErrorIsNil)
 	out := cmdtesting.Stdout(ctx)
 	c.Assert(out, gc.Equals, `
-ID         Scope  Revision  Backend  Path          Age
-666  application  2         juju     app.password  0001-01-01  
+ID   Revision  Rotate  Backend  Path          Age
+666         2  1 hour  juju     app.password  0001-01-01
```
Comment (hpidcock) [inline cmd/juju/secrets/list_test.go:66]: Age implies a duration, maybe we can use the new humanizeDuration or change the title.

File: worker/uniter/runner/jujuc/context.go [hunk 3]
```diff
@@ -163,13 +163,18 @@ type ContextUnit interface {
 	CloudSpec() (*params.CloudSpec, error)
 }
 
+type UpsertArgs struct {
+	Value          secrets.SecretValue
```
Comment (hpidcock) [inline worker/uniter/runner/jujuc/context.go:167]: Should these values themselves be optional?

**Review by hpidcock — APPROVED** [review]
Just a few questions, but LGTM

File: cmd/juju/secrets/list_test.go (hunk 2 above)
Comment (wallyworld) [reply cmd/juju/secrets/list_test.go:66]: Age is formatted with our existing `UserFriendlyDuration` func, which is for printing the difference between 2 times, one of them being `now`.
The Rotate value is an interval, every X hours etc, which is not quite the same.

File: cmd/juju/secrets/list.go (hunk 1 above)
Comment (wallyworld) [reply cmd/juju/secrets/list.go:165]: I can move to cmd/juju/common and add next to `UserFriendlyDuration`

File: worker/uniter/runner/jujuc/context.go (hunk 3 above)
Comment (wallyworld) [reply worker/uniter/runner/jujuc/context.go:167]: Value is an interface so nil supported.
With Rotate, a negative value was going to mean "unset". Easier than using pointers IMO. I'll add a comment

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$Merge$$

**Pull request merged**